### PR TITLE
feat: Github Codespaces Plugin

### PR DESCRIPTION
**Pull request by pyxelpioneer**
Fixes: amplication/amplication#7012

## PR Details

Adds the plugin as per defined in issue #7012

## PR Checklist

- [X] `npm build` doesn't throw any error
- [X] `npm test` doesn't throw any error
- [X] tested locally

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

**Comment by overbit** [comment]
@pyxelpioneer could you please address the CI failures?

File: plugins/devcontainer-github-codespaces/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    var containerConfig;
+    if (!settings.includeAdminUI) {
+      containerConfig = genDevcontainerConfig(serviceName, context.serverDirectories.baseDirectory)
+    } else {
+      containerConfig = genDevcontainerConfigWithAdminUI(serviceName, context.serverDirectories.baseDirectory, context.clientDirectories.baseDirectory)
+    }
+
+    const envConfig = fs.readFileSync(join(__dirname, "templates", "init.sh"), { encoding: "utf-8" })
+      .replace("{SERVER_ROOT}", context.serverDirectories.baseDirectory)
+      .replace("{CLIENT_ROOT}", context.clientDirectories.baseDirectory)
```
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/src/index.ts:54]: I don't think `CLIENT_ROOT` is there in init.sh

File: plugins/devcontainer-github-codespaces/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,6406 @@
+{
+  "name": "@amplication/devcontainer-github-codespaces",
+  "version": "1.0.0",
```
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/package-lock.json:3]: Please run `npm i` to update this.

File: plugins/devcontainer-github-codespaces/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    // Merge code
+    await modules.set({
+      code: JSON.stringify(containerConfig),
```
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/src/index.ts:58]: You can use prettier to prettify the code also.

image

Comment (Shurtu-gal) [reply plugins/devcontainer-github-codespaces/src/index.ts:58]: Example:- https://github.com/amplication/plugins/blob/e8b38999d71c29303e856a269acf77a6522e3957/plugins/formatter-prettier/src/index.ts#L63

**Review by Shurtu-gal — CHANGES_REQUESTED** [review]
Nice work 🚀. Just some tiny changes 💯

File: plugins/devcontainer-github-codespaces/src/static/admin.docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+version: "3.8"
+
+services:
+  admin-ui:
+    container_name: admin-ui-container
+    image: admin-ui-image
+    build:
+      context: ../admin-ui
```
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/src/static/admin.docker-compose.yml:8]:
```suggestion
      context: .
```

File: plugins/devcontainer-github-codespaces/src/tests/.keep
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/src/tests/.keep:1]: @pyxelpioneer Tests can be added for various cases of settings.

File: plugins/devcontainer-github-codespaces/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# @amplication/plugin-devcontainer-github-codespaces
+
+[![NPM Downloads](https://img.shields.io/npm/dt/@amplication/plugin-devcontainer-github-codespaces)](https://www.npmjs.com/package/@amplication/plugin-devcontainer-github-codespaces)
+
+Quickly generates the configuration that you require to setup devcontainer for your Amplication App
+
+## Purpose
+
+Generates the configuration that you require to setup devcontainer for your Amplication App
```
Comment (Shurtu-gal) [inline plugins/devcontainer-github-codespaces/README.md:9]: Duplicate of the earlier statement.

**Review by Shurtu-gal — CHANGES_REQUESTED** [review]
@pyxelpioneer Can you update the README to provide some more context as well as steps on how to startup the dev container after app is generated.

**Comment by Shurtu-gal** [comment]
> @pyxelpioneer Can you update the README to provide some more context as well as steps on how to startup the dev container after app is generated.

@pyxelpioneer What I meant was adding steps on how to run the dev container after the service is made.

**Review by Shurtu-gal — CHANGES_REQUESTED** [review]

**Comment by Shurtu-gal** [comment]
@overbit rest everything is fine by me. Could you have a look once too?

File: plugins/devcontainer-github-codespaces/src/tests/.keep
Comment (mulygottlieb) [inline plugins/devcontainer-github-codespaces/src/tests/.keep:1]: Why do we need this?

**Review by mulygottlieb — COMMENTED** [review]
Nice work @pyxelpioneer !
I added one comment about an empty file in the tests directory, but other than that it LGTM.

**Review by overbit — APPROVED** [review]

**Comment by levivannoort** [comment]
frok to branch

**Comment by levivannoort** [comment]
Hey @pyxelpioneer I just checkout the plugin, built it locally and generated the application + plugin code. 

I decided to try both without and with the admin and a remark there is that if I run without it the plugin will still generate a pointer to the compose file (for a non generated part of the application): 

with:
```
"dockerComposeFile": [
  "../app/server/docker-compose.yml",
  "../app/admin/docker-compose.yml"
],
```
without:
```
"dockerComposeFile": [
  "../app/server/docker-compose.yml",
  "../app/admin/docker-compose.yml"
],
```

I'd also suggest to change the name of the directory created within the `.devcontainer` to use snake-case so that there are no spacers in the name generated by the plugin.

I'll also try to run the generated code via the dev container ang get back to you on that.

**Review by overbit — CHANGES_REQUESTED** [review]
@pyxelpioneer can you check @levivannoort comments as we identified a bug with the adminui.

In addition, to simplify the approval and release process, would you mind adding a short video here showing the GitHub Codespaces in action with a simple amplication generated app?

**Comment by Shurtu-gal** [comment]
@pyxelpioneer any updates on this ?

**Comment by pyxelpioneer** [comment]
Hey @overbit @Shurtu-gal, sorry for the extreme delay due to shifting, my codespace limit gettting full etc. Thanks for not closing this PR!

Here is the video showing the output based on the rework of the plugin.


https://github.com/amplication/plugins/assets/151645921/1259d3cf-f0b1-4d10-9a6b-aee0f37a5417

**Comment by pyxelpioneer** [comment]
> Hey @overbit @Shurtu-gal @levivannoort , sorry for the extreme delay due to shifting, my codespace limit gettting full etc. Thanks for not closing this PR!
> 
> Here is the video showing the output based on the rework of the plugin.
> dance-2024-04-07_22.online-video-cutter.com.mp4

Hey @overbit ?

**Comment by overbit** [comment]
Well done @pyxelpioneer! We are reviewing it.

**Review by overbit — APPROVED** [review]

**Comment by pyxelpioneer** [comment]
> Well done @pyxelpioneer! We are reviewing it.

Thanks :heart_eyes:

**Review by Shurtu-gal — APPROVED** [review]

**Comment by Shurtu-gal** [comment]
Great Work @pyxelpioneer!!!

**Pull request merged**